websocket/session.go

```diff
@@ -0,0 +1,325 @@
+
+	// Cancel context
+	s.ctxCancel()
+
+	// Close connection
+	s.conn.Close()
+	s.conn = nil
+}
+
+// sendPresence sends user presence: "connected", "disconnected" and "" activity kinds
+func (s *session) sendPresence(kind string) error {
+	//connections := store.CountConnections(s.user.Identity())
+	//if kind == "disconnected" {
+	//	connections--
+	//}
+
+	return nil
+}
+
+func (s *session) Handle() (err error) {
+	if err = s.connected(); err != nil {
+		s.Close()
+		return
+	}
+
+	go func() {
+		_ = s.readLoop()
+	}()
+	return s.writeLoop()
+}
+
+func (s *session) Close() {
+	s.once.Do(func() {
+		s.disconnected()
+		s.Delete()
+	})
+}
+
+func (s *session) readLoop() (err error) {
+	defer func() {
+		s.Close()
+	}()
+
+	if err = s.conn.SetReadDeadline(time.Now().Add(s.config.PingTimeout)); err != nil {
+		return
+	}
+
+	s.conn.SetPongHandler(func(string) error {
+		return s.conn.SetReadDeadline(time.Now().Add(s.config.PingTimeout))
+	})
+
+	s.remoteAddr = s.conn.RemoteAddr().String()
+
+	for {
+		_, raw, err := s.conn.ReadMessage()
+		if err != nil {
+			return errors.Wrap(err, "s.readLoop")
+		}
+
+		if err = s.dispatch(raw); err != nil {
+			s.log(zap.Error(err)).Error("could not dispatch")
+			//_ = s.send(outgoing.NewError(err))
+		}
+	}
+}
+
+func (s *session) writeLoop() error {
+	ticker := time.NewTicker(s.config.PingPeriod)
+
+	defer func() {
+		ticker.Stop()
+		s.Close() // break readLoop
+	}()
+
+	write := func(msg []byte) (err error) {
+		if s.conn == nil {
+			// Connection closed, nowhere to write
+			return
+		}
+
+		if err = s.conn.SetWriteDeadline(time.Now().Add(s.config.Timeout)); err != nil {
+			return
+		}
+
+		if msg != nil && s.conn != nil {
+			return s.conn.WriteMessage(gWebsocket.TextMessage, msg)
+		}
+
+		return
+	}
+
+	ping := func() (err error) {
+		if s.conn == nil {
+			// Connection closed, nothing to ping
+			return
+		}
+
+		if err = s.conn.SetWriteDeadline(time.Now().Add(s.config.Timeout)); err != nil {
+			return
+		}
+
+		if s.conn != nil {
+			return s.conn.WriteMessage(gWebsocket.PingMessage, nil)
+		}
+
+		return
+	}
+
+	for {
+		select {
+		case msg, ok := <-s.send:
+			if !ok {
+				// channel closed
+				return nil
+			}
+
+			if err := write(msg); err != nil {
+				return errors.Wrap(err, "writeLoop send")
+			}
+
+		case msg := <-s.stop:
+			// Shutdown requested, don't care if the message is delivered
+			_ = write(msg)
+			return nil
+
+		case <-ticker.C:
+			if err := ping(); err != nil {
+				return errors.Wrap(err, "writeLoop ping")
+			}
+		}
+	}
+}
+
+func (s *session) dispatch(raw []byte) error {
+	var p, err = Unmarshal(raw)
+	if err != nil {
+		return errors.Wrap(err, "Session.incoming: payload malformed")
+	}
+
+	ctx := s.Context()
+
+	if p.Auth != nil {
+		return s.authenticate(ctx, p.Auth)
+	}
+
+	return nil
+}
+
+func (s *session) Save() *session {
+	if s.id == 0 {
+		s.id = id.Next()
+	}
+
+	if s.user != nil {
+		if _, ok := sessions[s.user.Identity()]; !ok {
```

Comment:
so... one session per user? will this work?

Reply:
No, Fixed it!